[SPARK-14163][CORE] SumEvaluator and countApprox cannot reliably handle RDDs of size 1

mtustin-handy opened the pull request:
## What changes were proposed in this pull request?

This special cases 0 and 1 counts to avoid passing 0 degrees of freedom. 
## How was this patch tested?

Tests run successfully.

mtustin-handy committed:
Use confidence when count is 1.

Work with me in New York? https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala:
```diff
@@ -56,9 +56,12 @@ private[spark] class SumEvaluator(totalOutputs: Int, confidence: Double)
           new TDistribution(degreesOfFreedom).inverseCumulativeProbability(1 - (1 - confidence) / 2)
+        } else {
+          // this may not be statistically meaningful
+          confidence
```
This isn't valid; confidence is really a probability and confFactor is a number of standard deviations.

mtustin-handy replied:
Right, Double.PositiveInfinity it is then.

mtustin-handy replied:
Updated.

mtustin-handy committed:
Set infinite confidence if count is 1 or otherwise confidence would be NaN.
Add test + equality methods for BoundedDouble

Work with me in NYC: https://www.handy.com/careers/73115?gh_jid=73115&gh_src=o5qcxn

srowen commented on core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala:
```diff
@@ -56,9 +56,12 @@ private[spark] class SumEvaluator(totalOutputs: Int, confidence: Double)
           new TDistribution(degreesOfFreedom).inverseCumulativeProbability(1 - (1 - confidence) / 2)
+        } else {
+          // No way to meaningfully estimate confidence, so we signal no particular confidence interval
+          Double.PositiveInfinity
```
Still don't think this works since you're multiplying infinity by NaN to get the intervals. It's too odd to begin with but results in NaN anyway.

mtustin-handy replied:
I'm not sure I follow you. If `sumStdev` is `NaN` it really won't matter what this value is. If it's any other value, multiplying by `Infinity` results in `Infinity`, which I believe is what is desired. 

`sumStdDev` will only be `NaN` if `sumVar` is negative. By the construction of `sumVar`, I don't think that's possible. 

```
scala> Double.PositiveInfinity * 2
res1: Double = Infinity

scala> Double.PositiveInfinity * 2.0
res2: Double = Infinity

scala> Double.PositiveInfinity * Double.PositiveInfinity
res3: Double = Infinity

scala> Double.PositiveInfinity * Double.NegativeInfinity
res4: Double = -Infinity

```

srowen replied:
But the relevant computation is:
```
scala> Double.NaN * Double.PositiveInfinity
res1: Double = NaN
```

The interval should not be `NaN`.

If count is 1 then the sample variance is NaN along with everything else that's a function of a second moment.

srowen replied:
Backing up, this change should probably include a test for this reason, to verify the behavior.

mtustin-handy replied:
Re: sampleVariance, that's the missing piece I didn't understand. Re: tests, I'm figuring out now how to run individual tests and where they would be if they exist.

mtustin-handy replied:
Look out for updates to this tomorrow (Eastern Time; so probably we'll loop back on this Tuesday).

rxin commented:
Do you guys use this feature? I'm actually thinking we should deprecate this feature over time.

srowen commented:
@rxin these are used to implement `sumApprox`, `meanApprox`, etc on `RDD`. I don't have a view on whether those are used much. I figure we can go ahead and fix it for this case right now in any event.

rxin commented:
Yea fixing them is fine. I'm asking because this is the 1st time I've seen people using this api in public and want to get some feedback here.

mtustin-handy commented:
I hit this with `countApprox`.

mtustin-handy commented:
@rxin @srowen Looks like github is not showing the correct diff any more. I'm going to close this, reopen and tag you.

srowen commented:
@mtustin-handy you can just push new commits to the branch.
